newm-server/src/main/kotlin/io/newm/server/features/song/AgreementMessageReceiver.kt

```diff
@@ -0,0 +1,45 @@
+package io.newm.server.features.song
+
+import com.amazonaws.services.sqs.model.Message
+import io.ktor.server.application.*
+import io.ktor.util.logging.*
+import io.newm.server.aws.SqsMessageReceiver
+import io.newm.server.di.inject
+import io.newm.server.ext.toUUID
+import io.newm.server.features.song.model.AgreementMessage
+import io.newm.server.features.song.model.MintingStatus
+import io.newm.server.features.song.model.Song
+import io.newm.server.features.song.repo.SongRepository
+import kotlinx.coroutines.runBlocking
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import org.slf4j.MarkerFactory
+
+class AgreementMessageReceiver : SqsMessageReceiver {
+    private val repository: SongRepository by inject()
+    private val logger: Logger by inject()
+    private val marker = MarkerFactory.getMarker(javaClass.simpleName)
+
+    @OptIn(ExperimentalSerializationApi::class)
+    private val json = Json {
+        ignoreUnknownKeys = true
+        explicitNulls = false
+    }
+
+    override fun onMessageReceived(message: Message) {
+        val msg: AgreementMessage = json.decodeFromString(message.body)
+        logger.debug(marker, "Agreement job config ID: ${msg.configurationId}")
+
+        if (msg.configurationId != "agreement-created") return
+        val key = msg.key ?: return
+
+        val songId = key
+            .substringBefore('/')
+            .toUUID()
+
+        runBlocking {
+            repository.update(songId, Song(mintingStatus = MintingStatus.StreamTokenAgreementApproved))
```

Comment:
What happens if we fail to update the database? Does the system retry to re-deliver the SQS message to another receiver? Does it end up in a dead-letter queue?

Reply:
Currently, only one of the receivers will be able to consume and delete the message from the queue. If the database update fails for some reason (e.g., Song was deleted), then the message is ignored.
Amazon SQS supports dead-letter queues: https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-sqs-dead-letter-queues.html
Perhaps, we could configure it in such a way that on failure we move the message to an SQS dead-letter queue instead of delete it.